### PR TITLE
Team jobs meta-schema (values-schema.yaml)

### PR DESCRIPTION
**Pull request by svatwork**
This is an iteration on #239. 

I think the values schema for team jobs is more complete now, but please give me some criticism because I must have left out some features.

### CODE REVIEW

File: values-schema.yaml [hunk 1]
```diff
@@ -207,6 +221,60 @@ definitions:
...
+        type: string
+      volumes:
+        $ref: '#/definitions/volumes'
+  jobs:
+    description: You may pass an array of Job objects. The Kubernetes API accepts
+      a Job specification, which is meant for run-to-completion apps (e.g. batch
+      Jobs). Job names must be globally unique.
+    items:
+      $ref: '#/definitions/job'
+    type: array
```
Comment (svatwork) [inline values-schema.yaml:277]: I've used all configurable keys from job.yaml, is this correct?

File: .demo/env/clouds/aws/demo/jobs.demo.yaml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+teamConfig:
+    teams:
+        demo:
+            jobs:
+                - name: job-hello-world
+                  enabled: true
```
Comment (Morriz) [inline .demo/env/clouds/aws/demo/jobs.demo.yaml:6]: this should be default and then we drop the flag

Comment (Morriz) [reply .demo/env/clouds/aws/demo/jobs.demo.yaml:6]: and it seems to make no sense to just run it on deployment imo, cause that is what this will do.

I think it makes sense to look at the options helm offers and maybe expose that here as configuration. Like always deleting and starting a job fresh, or only when the spec changed. I am not sure these options exist but please investigate what is possible for our users.

File: values-schema.yaml [hunk 3]
```diff
@@ -172,8 +172,22 @@ definitions:
...
+        name:
+          pattern: '^[a-zA-Z0-9_]*$'
+          type: string
+        value:
+          maxLength: 131072
+          type: string
+      required:
+        - name
+        - value
+      type: object
```
Comment (Morriz) [inline values-schema.yaml:187]: please don't use verbose version and allow `key: value` format for easier input

File: values-schema.yaml [hunk 4]
```diff
@@ -207,6 +221,68 @@ definitions:
...
+          script:
+            $ref: '#/definitions/script'
+          shell:
+            pattern: (sh|bash)
+      mounts:
+        $ref: '#/definitions/mounts'
+      name:
+        default: job-hello-world
+        description: |
+          Must be a string starting with 'job-'. Name must be globally unique.
```
Comment (Morriz) [inline values-schema.yaml:248]: why is that made mandatory? It does not need to be

File: values-schema.yaml [hunk 5]
```diff
@@ -207,6 +221,68 @@ definitions:
...
+          May be a non-empty string containing the name of the serviceAccount
+          that is authorized to launch the containers that will execute the job. 
+          Defaults to 'default' Service Account.
+        type: string
+      ttlSecondsAfterFinished:
+        pattern: ^[0-9]*$
+        type: number
+      type:
+        default: job
+        pattern: (job|cronjob)
```
Comment (Morriz) [inline values-schema.yaml:268]: please use enum

File: values-schema.yaml [hunk 6]
```diff
@@ -207,6 +221,68 @@ definitions:
...
+      - enabled
+      - image
+      - name
+      - script
+      - type
+
+  jobs:
+    description: You may pass an array of Job objects. The Kubernetes API accepts
+      a Job specification, which is meant for run-to-completion apps (e.g. batch
+      Jobs). Job names must be globally unique.
```
Comment (Morriz) [inline values-schema.yaml:282]: globally? really? I think it needs to be unique in the namespace

File: values-schema.yaml [hunk 7]
```diff
@@ -274,6 +365,10 @@ definitions:
         type: number
       minReplicas:
         type: number
+  script:
+    default: echo "Hello world"
```
Comment (Morriz) [inline values-schema.yaml:369]: why have a default here? it should fail when empty imo...remove the default

File: values-schema.yaml [hunk 8]
```diff
@@ -528,6 +612,21 @@ definitions:
   url:
     pattern: '^(https:\/\/)([\w\-])+\.{1}([a-zA-Z]{2,63})([\/\w-]*)*\/?\??([^#\n\r]*)?#?([^\n\r]*)$'
     type: string
+  volume:
+    additionalProperties: false
+    properties:
+      name:
+        description: Must be a string starting with 'volume-'. Name must be globally unique.
```
Comment (Morriz) [inline values-schema.yaml:619]: again, not true...not needed

**Review by Morriz — CHANGES_REQUESTED** [review]

File: values-schema.yaml (hunk 3 above)
Comment (svatwork) [reply values-schema.yaml:187]: I just copied this block to a definition to make it more DRY, but I can definitely do it. So you want to include a refactor of this inside the scope?

File: values-schema.yaml (hunk 6 above)
Comment (svatwork) [reply values-schema.yaml:282]: Sounds good, I now think that sentence doesn't makes sense here...

File: .demo/env/clouds/aws/demo/jobs.demo.yaml (hunk 2 above)
Comment (svatwork) [reply .demo/env/clouds/aws/demo/jobs.demo.yaml:6]: > this should be default and then we drop the flag

If I understand correctly, if we add this to the scope then we should add that to the chart. 

> and it seems to make no sense to just run it on deployment imo, cause that is what this will do.

I think it makes sense to look at the options helm offers and maybe expose that here as configuration. Like always deleting and starting a job fresh, or only when the spec changed. I am not sure these options exist but please investigate what is possible for our users.

I'll add that to the scope as well. It makes sense to me to explore that yes.

Comment (svatwork) [reply .demo/env/clouds/aws/demo/jobs.demo.yaml:6]: If so, I've updated the [spec](https://github.com/redkubes/otomi-core/issues/204), please verify if this is the intended scope.

File: values-schema.yaml (hunk 7 above)
Comment (j-zimnowoda) [inline values-schema.yaml]: It should not have a default value as container image can have default executable command that we are not aware of.

File: values-schema.yaml [hunk 9]
```diff
@@ -207,6 +221,68 @@ definitions:
...
+        $ref: '#/definitions/image'
+      init:
+        properties:
+          image:
+            $ref: '#/definitions/image'
+          script:
+            $ref: '#/definitions/script'
+          shell:
+            pattern: (sh|bash)
+      mounts:
```
Comment (j-zimnowoda) [inline values-schema.yaml:244]: I am wondering, how a Otomi-console user can use it?

**Review by j-zimnowoda — CHANGES_REQUESTED** [review]

File: values-schema.yaml (hunk 9 above)
Comment (svatwork) [reply values-schema.yaml:245]: > I am wondering, how a Otomi-console user can use it?

As we spoke about in the chat, @j-zimnowoda is talking about the `mounts` key. If I understand correctly, @j-zimnowoda thinks it might be confusing to not expose mounts to the otomi-console but still allow users to define it in values. 

So I guess that implies we don't expose mounts by default. And if it's not by default, then the property also isn't mandatory in the values schema (@j-zimnowoda is this correct?)

What is your input on this @Morriz? I have just defined it here because there's a property in the Job chart that would match also having a definition in the values schema. I didn't think it would have any other implications other than giving the users the option to define mounts for their jobs, I can't imagine it wouldn't be useful.

File: .demo/env/clouds/aws/demo/jobs.demo.yaml (hunk 2 above)
Comment (svatwork) [reply .demo/env/clouds/aws/demo/jobs.demo.yaml:6]: > and it seems to make no sense to just run it on deployment imo, cause that is what this will do.

_Note: I am exploring options to run it on spec change._

I just ran the command from the perspective of an end-user; it will run on deployment, yes. But having made no changes to the spec, it will still run a new one (see output): 

```
otomi apply -l name=jobs-demo
Comparing release=jobs-demo, chart=../charts/jobs
	Release was not present in Helm.  Diff will show entire contents as new.
team-demo, 0-888cdb3c-5da2-4c50-9779-9b45f689a053, Job (batch) has been added:
-
+ # Source: jobs/templates/job.yaml
+ apiVersion: batch/v1
+ kind: Job
+ metadata:
+   name: 0-888cdb3c-5da2-4c50-9779-9b45f689a053
+   labels:
+     helm.sh/chart: jobs-0.1.0
+     app.kubernetes.io/name: jobs
+     app.kubernetes.io/instance: jobs-demo
+     app.kubernetes.io/version: "0.1.0"
+     app.kubernetes.io/managed-by: Helm
+ spec:
+   template:
+     metadata:
+       labels:
+         helm.sh/chart: jobs-0.1.0
+         app.kubernetes.io/name: jobs
+         app.kubernetes.io/instance: jobs-demo
+         app.kubernetes.io/version: "0.1.0"
+         app.kubernetes.io/managed-by: Helm
+       annotations:
+         checksum/secret: 01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546
+         checksum/config: 01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546
+     spec:
+       serviceAccountName: default
+       containers:
+       - image: "busybox:latest"
+         imagePullPolicy: "IfNotPresent"
+         name: "0"
+         command: ["sh", "-c"]
+         args:
+           - |
+             echo "Hello World"
+         resources:
+           limits:
+             cpu: 200m
+             memory: 256Mi
+           requests:
+             cpu: 100m
+             memory: 128Mi
+         envFrom:
+         volumeMounts:
+           - name: fakeroot
+             mountPath: /fakeroot.pem
+             subPath: fakeroot.pem
+       restartPolicy: Never
+       volumes:
+         - name: fakeroot
+           configMap:
+             name: fakeroot
+   backoffLimit: 3
+   ttlSecondsAfterFinished: 86400
Upgrading release=jobs-demo, chart=../charts/jobs
Release "jobs-demo" does not exist. Installing it now.
NAME: jobs-demo
LAST DEPLOYED: Tue Dec 22 11:25:04 2020
NAMESPACE: team-demo
STATUS: deployed
REVISION: 1
TEST SUITE: None
Listing releases matching ^jobs-demo$
jobs-demo	team-demo	1       	2020-12-22 11:25:04.762810766 +0000 UTC	deployed	jobs-0.1.0	0.1.0
UPDATED RELEASES:
NAME        CHART            VERSION
jobs-demo   ../charts/jobs     0.1.0
 sebastiaanverbeek@mbprosv  ~/Projects/redkubes/dev/otomi-core   task/204/spec-enabled ± otomi apply -l name=jobs-demo
Comparing release=jobs-demo, chart=../charts/jobs
team-demo, 0-e4c9d60c-8522-4a1f-b5d6-efc9a0a20cd4, Job (batch) has been removed:
- # Source: jobs/templates/job.yaml
- apiVersion: batch/v1
- kind: Job
- metadata:
-   name: 0-e4c9d60c-8522-4a1f-b5d6-efc9a0a20cd4
-   labels:
-     helm.sh/chart: jobs-0.1.0
-     app.kubernetes.io/name: jobs
-     app.kubernetes.io/instance: jobs-demo
-     app.kubernetes.io/version: "0.1.0"
-     app.kubernetes.io/managed-by: Helm
- spec:
-   template:
-     metadata:
-       labels:
-         helm.sh/chart: jobs-0.1.0
-         app.kubernetes.io/name: jobs
-         app.kubernetes.io/instance: jobs-demo
-         app.kubernetes.io/version: "0.1.0"
-         app.kubernetes.io/managed-by: Helm
-       annotations:
-         checksum/secret: 01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546
-         checksum/config: 01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546
-     spec:
-       serviceAccountName: default
-       containers:
-       - image: "busybox:latest"
-         imagePullPolicy: "IfNotPresent"
-         name: "0"
-         command: ["sh", "-c"]
-         args:
-           - |
-             echo "Hello World"
-         resources:
-           limits:
-             cpu: 200m
-             memory: 256Mi
-           requests:
-             cpu: 100m
-             memory: 128Mi
-         envFrom:
-         volumeMounts:
-           - name: fakeroot
-             mountPath: /fakeroot.pem
-             subPath: fakeroot.pem
-       restartPolicy: Never
-       volumes:
-         - name: fakeroot
-           configMap:
-             name: fakeroot
-   backoffLimit: 3
-   ttlSecondsAfterFinished: 86400
+
team-demo, 0-fe85431a-1f1a-422d-bf44-f6fe58f53628, Job (batch) has been added:
-
+ # Source: jobs/templates/job.yaml
+ apiVersion: batch/v1
+ kind: Job
+ metadata:
+   name: 0-fe85431a-1f1a-422d-bf44-f6fe58f53628
+   labels:
+     helm.sh/chart: jobs-0.1.0
+     app.kubernetes.io/name: jobs
+     app.kubernetes.io/instance: jobs-demo
+     app.kubernetes.io/version: "0.1.0"
+     app.kubernetes.io/managed-by: Helm
+ spec:
+   template:
+     metadata:
+       labels:
+         helm.sh/chart: jobs-0.1.0
+         app.kubernetes.io/name: jobs
+         app.kubernetes.io/instance: jobs-demo
+         app.kubernetes.io/version: "0.1.0"
+         app.kubernetes.io/managed-by: Helm
+       annotations:
+         checksum/secret: 01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546
+         checksum/config: 01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546
+     spec:
+       serviceAccountName: default
+       containers:
+       - image: "busybox:latest"
+         imagePullPolicy: "IfNotPresent"
+         name: "0"
+         command: ["sh", "-c"]
+         args:
+           - |
+             echo "Hello World"
+         resources:
+           limits:
+             cpu: 200m
+             memory: 256Mi
+           requests:
+             cpu: 100m
+             memory: 128Mi
+         envFrom:
+         volumeMounts:
+           - name: fakeroot
+             mountPath: /fakeroot.pem
+             subPath: fakeroot.pem
+       restartPolicy: Never
+       volumes:
+         - name: fakeroot
+           configMap:
+             name: fakeroot
+   backoffLimit: 3
+   ttlSecondsAfterFinished: 86400
Upgrading release=jobs-demo, chart=../charts/jobs
Release "jobs-demo" has been upgraded. Happy Helming!
NAME: jobs-demo
LAST DEPLOYED: Tue Dec 22 11:27:28 2020
NAMESPACE: team-demo
STATUS: deployed
REVISION: 2
TEST SUITE: None
Listing releases matching ^jobs-demo$
jobs-demo	team-demo	2       	2020-12-22 11:27:28.365770446 +0000 UTC	deployed	jobs-0.1.0	0.1.0
UPDATED RELEASES:
NAME        CHART            VERSION
jobs-demo   ../charts/jobs     0.1.0
```

What do you mean by making "no sense to just run it on deployment"? It's obvious there are some use-cases for this. but if I understand correctly, this already covers "always deleting and starting a job fresh" or "when the spec changed".

File: values-schema.yaml (hunk 3 above)
Comment (Morriz) [reply values-schema.yaml:188]: Again, please refactor this to use `key: value` format using `patternProperties`, as we allow all env vars (or anything actually) to be specified in a DRY fashion

File: values-schema.yaml (hunk 9 above)
Comment (Morriz) [reply values-schema.yaml:245]: I agree, it is confusing. So far we let them mount configmaps and secrets already, but that might not be enough.

Use case: user wants a config file mounted. How can they? I think we should quickly allow for users to add configmaps as well, so that they can mount those. Please add that to this PR

File: values-schema.yaml [hunk 10]
```diff
@@ -346,6 +444,12 @@ definitions:
     required:
       - name
       - type
+  shell:
+    default: sh
+    enum:
+      - sh
+      - bash
```
Comment (Morriz) [inline values-schema.yaml:451]: why is this needed? it is encapsulated by the running script. otomi-core should not know about this.

File: values-schema.yaml [hunk 11]
```diff
@@ -528,6 +621,20 @@ definitions:
...
+    additionalProperties: false
+    properties:
+      name:
+        description: Name must match mount name.
+        type: string
+      configMap:
+        properties:
+          name:
+            type: string
+  volumes:
```
Comment (Morriz) [inline values-schema.yaml:634]: both `volume` and `volumes`? Only the latter please

**Review by Morriz — CHANGES_REQUESTED** [review]

File: .demo/env/clouds/aws/demo/jobs.demo.yaml (hunk 2 above)
Comment (svatwork) [reply .demo/env/clouds/aws/demo/jobs.demo.yaml:6]: @Morriz Would you please take a look at this, then I can perform the requested changes and finish the PR.

Comment (Morriz) [reply .demo/env/clouds/aws/demo/jobs.demo.yaml:6]: don't start work without requirements please...as always ;)

We want predictable outcomes for the following use cases:

- cronjob (easy, covered)
- always run upon deployment (can be realized by removing the job after it finished succesfuly)
- only run upon deployment when job changed (might need to realize it to "never remove", possibly with expiry of zero, but you have to assert that)

I don't think we need to write tests for these use cases as we would test for k8s doing its work

File: values-schema.yaml (hunk 3 above)
Comment (Morriz) [reply values-schema.yaml:188]: @svatwork has this been done?

File: .demo/env/clouds/aws/demo/jobs.demo.yaml (hunk 2 above)
Comment (Morriz) [reply .demo/env/clouds/aws/demo/jobs.demo.yaml:6]: So again: the `enabled` flag has nothing to do with this and should be removed

File: values-schema.yaml [hunk 12]
```diff
@@ -207,6 +223,70 @@ definitions:
...
+        $ref: '#/definitions/resources'
+      schedule:
+        description: Must give a cron-type expression if the job-type is 'cronjob' (job.type).
+        type: string
+      script:
+        $ref: '#/definitions/script'
+      secret:
+        $ref: '#/definitions/secret'
+      serviceAccountName:
+        default: default
```
Comment (Morriz) [inline values-schema.yaml:262]: we don't allow setting this, so please remove

**Review by Morriz — CHANGES_REQUESTED** [review]

File: .demo/env/clouds/aws/demo/jobs.demo.yaml (hunk 2 above)
Comment (svatwork) [reply .demo/env/clouds/aws/demo/jobs.demo.yaml:6]: > don't start work without requirements please...as always ;)

I haven't started anything, what do you mean exactly?

> always run upon deployment (can be realized by removing the job after it finished succesfuly)

Following your thoughts, that is what happens with the above command output. 

> only run upon deployment when job changed (might need to realize it to "never remove", possibly with expiry of zero, but you have to assert that)

> I don't think we need to write tests for these use cases as we would test for k8s doing its work

Again following your thoughts, you imply this case isn't covered right now, is this correct? Do you suggest to have at least a key where an user can specify what the deployment strategy will be? Something like:

```
deployment: always|onChange
```

The research I've done is that a job should only run once and not have any logic for re-running a job. It will be only possible by duplicating the old spec and rerunning it with a hacky kubectl bash command. I think it would be best to look into helm hooks.

> So again: the enabled flag has nothing to do with this and should be removed

You don't have to repeat yourself, but the fault is mine this time. The production ready version will be #258, unless it will be best to merge these branches. After this discussion I intend to close this pull request. I've already made the change there.

Comment (svatwork) [reply .demo/env/clouds/aws/demo/jobs.demo.yaml:6]: I'll merge values-schema.yaml into this branch so we can agree on that after review.

File: values-schema.yaml (hunk 10 above)
Comment (svatwork) [reply values-schema.yaml:433]: Right now it's encapsulated explicitly: `command: ["{{ $spec.init.shell | default "sh" }}", "-c"]`. Do you suggest to remove this reference?

File: values-schema.yaml (hunk 9 above)
Comment (svatwork) [reply values-schema.yaml:237]: I have defined some behaviour for it under the `volumes` definition in this `values-schema.yaml`: 

```volumes:
    items:
      additionalProperties: false
      properties:
        name:
          description: Name must match mount name.
          type: string
        configMap:
          properties:
            name:
              type: string
```

Is this enough? Right now it's only in line with the `otomi-core` chart.

File: values-schema.yaml [hunk 13]
```diff
@@ -63,6 +63,7 @@ definitions:
     $ref: '#/definitions/labelsAnnotations'
     description: A set of annotations.
   azureMonitor:
+    readOnly: false
```
Comment (Morriz) [inline values-schema.yaml:66]: Why add this prop? What is it for? Is it needed?

File: values-schema.yaml [hunk 14]
```diff
@@ -207,6 +216,55 @@ definitions:
       tag:
         type: string
     title: Container image
+    type: object
```
Comment (Morriz) [inline values-schema.yaml:219]: Redundant

File: values-schema.yaml [hunk 15]
```diff
@@ -207,6 +216,55 @@ definitions:
...
+          - cronjob
+        type: string
+      volumes:
+        $ref: '#/definitions/volumes'
+    required:
+      - image
+      - script
+      - type
+  jobs:
+    description: You may pass an array of Job objects. The Kubernetes API accepts
```
Comment (Morriz) [inline values-schema.yaml:261]: Would you mind adding a period to every description now that you started putting them at the end of yours? ;)

File: values-schema.yaml [hunk 16]
```diff
@@ -937,13 +1014,6 @@ properties:
                 $ref: '#/definitions/resources'
               pilot:
                 $ref: '#/definitions/resources'
-      jobs:
```
Comment (Morriz) [inline values-schema.yaml:940]: These previous jobs need their own place. These jobs are hard named so maybe put them under "tasks". So it corresponds with `otomi-tasks`.

**Review by Morriz — CHANGES_REQUESTED** [review]

File: values-schema.yaml (hunk 15 above)
Comment (svatwork) [reply values-schema.yaml:261]: I can also remove mine?

File: values-schema.yaml (hunk 16 above)
Comment (svatwork) [reply values-schema.yaml:940]: Makes sense, will do.

File: values-schema.yaml (hunk 15 above)
Comment (Morriz) [reply values-schema.yaml:261]: When we have multiline descriptions it makes sense to use periods.

File: values-schema.yaml (hunk 10 above)
Comment (Morriz) [reply values-schema.yaml]: We don't want to know about what to run or how. Expect the tooling to be in the container and just use the default shell. No choosing here

Comment (Morriz) [reply values-schema.yaml]: if they want to start another type of shell they can if they built it so

File: values-schema.yaml (hunk 16 above)
Comment (svatwork) [reply values-schema.yaml:940]: Well actually, just putting them back doesn't break anything. I don't know why I deleted them in the first place.

Comment (Morriz) [reply values-schema.yaml:940]: That is fine indeed.
